sw/device/lib/dif/README.md

```diff
@@ -50,7 +50,144 @@ memory-mapped hardware, and modifying processor CSRs.
+  does not perform any hardware operations. `handle` may point to uninitialized
+  data but must point to initialized data if this function returns `Ok`.
+* `result_t dif_<ip>_configure(const dif_<ip>_t *handle, dif_<ip>_config_t
+  config);` is a function that configures the hardware managed by `handle` with
+  runtime parameters in an implementation-defined way. This function should be
+  "one-off": it should only need to be called once for the lifetime of the handle.
+
+  If there is no meaningful state to configure, this function may be omitted.
+
+#### Interrupt Management
+
+The following types and functions are a standard inteface for peripherals that
+provide a collection of `INTR_ENABLE`, `INTR_STATE`, and `INTR_TEST` registers
+for interrupt management. A DIF library for a peripheral providing such
+registers must provide this interface.
+
+* `dif_<ip>_irq_t` is an enum that lists all of the interrupt types for this
+  peripheral. A DIF library may opt to use another pre-existing type instead,
+  when, for example, interrupt types are coupled to a distinct
+  peripheral-specific concept. In that case, all occurences of `dif_<ip>_irq_t`
+  belkow should be replaced with this typoe.
+* `result_t dif_<ip>_irq_is_pending(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type, bool *is_pending);` checks whether a specific interrupt is
+  pending (i.e., if the interrupt has been asserted but not yet cleared).
+* `result_t dif_<ip>_irq_clear(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type);` clears a specific interrupt (and does nothing if the interrupt is
+  not currrently pending).
+* `result_t dif_<ip>_irq_get_enabled(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type, dif_<ip>_toggle_t *state);` gets whether an interrupt is enabled
+  (i.e., masked).
+* `result_t dif_<ip>_irq_set_enabled(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type, dif_<ip>_toggle_t state);` sets whether a particular interrupt is
+  enabled (i.e., masked).
+* `result_t dif_<ip>_irq_force(const dif_<ip>_t *handle, dif_<ip>_irq_t
+  irq_type);` forcibly asserts a specific interrupt, causing it to be serviced
+  as if hardware had triggered it.
+
+Additionally, the following types allow for batch save/restore operations on
+the interrupt enable register:
+
+* `dif_<ip>_irq_snapshot_t` is a type that encapsulates restorable interrupt
+  state, to be used with the two functions below. This type should be treated as
+  opaque by clients (though it will usually just be a `uint32_t`).
+* `result_t dif_<ip>_irq_disable_all(const dif_<ip>_t *handle,
+  dif_<ip>_irq_snapshot_t *snapshot);` disables all interrupts associated with
+  the peripheral, saving them to `snapshot`. `snapshot` may be null, in which
+  case the previous enablement state is not saved.
+* `result_t dif_<ip>_irq_restore_all(const dif_<ip>_t *handle,
+  dif_<ip>_irq_snapshot_t snapshot);` restores an interrupt enablement snapshot
```

Comment:
Question: For symmetry, what's the feeling on this taking a `dif_<ip>_irq_snapshot_t *snapshot` instead?

Reply:
That's actually not a bad idea, since while a snapshot must be of fixed sized, it might be quite large.